### PR TITLE
fix unarchive integration tests when LC_ALL is not set to en_US.UTF-8

### LINKED ISSUE #58420: The unarchive integration test fails when LC_ALL is not set.
##### SUMMARY

The unarchive integration test fails when LC_ALL is not set.

##### ISSUE TYPE

Bug Report

##### COMPONENT NAME

unarchive

##### ANSIBLE VERSION

devel

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

1. Set `ansible_connection=ssh ansible_host=localhost` for `testhost` to prevent modules from inheriting the environment from the controller: https://github.com/ansible/ansible/blob/edcb5b6775adac09eee5412b499b8f0aa5194610/test/integration/inventory#L6
2. Run the test in an environment where LC_ALL is not set by default: `ansible-test integration unarchive --remote freebsd/12.0 -v`

##### EXPECTED RESULTS

Tests pass.

##### ACTUAL RESULTS

Tests traceback:

```
TASK [unarchive : test that unarchive works with an archive that contains non-ascii filenames] **************************************************************************
fatal: [testhost]: FAILED! => {"changed": false, "module_stderr": "Shared connection to localhost closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1561576726.3763218-37161750934991/AnsiballZ_unarchive.py\", line 139, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1561576726.3763218-37161750934991/AnsiballZ_unarchive.py\", line 131, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1561576726.3763218-37161750934991/AnsiballZ_unarchive.py\", line 65, in invoke_module\r\n    spec.loader.exec_module(module)\r\n  File \"<frozen importlib._bootstrap_external>\", line 678, in exec_module\r\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\r\n  File \"/tmp/ansible_unarchive_payload_o9_x930g/__main__.py\", line 909, in <module>\r\n  File \"/tmp/ansible_unarchive_payload_o9_x930g/__main__.py\", line 859, in main\r\n  File \"/usr/local/lib/python3.6/genericpath.py\", line 42, in isdir\r\n    st = os.stat(s)\r\nUnicodeEncodeError: 'ascii' codec can't encode characters in position 51-54: ordinal not in range(128)\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Failure on Shippable: https://app.shippable.com/github/ansible/ansible/runs/129521/18/tests

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #58420

Ensure run_command and isdir get bytes.

##### ISSUE TYPE
- Bugfix Pull Request

### CODE REVIEW

**Comment by sivel** [comment]
Overall this looks good.  But I wonder if we could just do the conversion to `to_bytes` in `main` and then pass that in, still maintaining the `b_dest` naming in the function signatures of the different classes?

I can see it both ways, currently the code doesn't assume context, it ensures it, but we then have more places to ensure we convert it, opening the door for future omissions.

**Review by mattclay — APPROVED** [review]

**Comment by s-hertel** [comment]
@sivel Is 8a669d36207eb61e3d769d8a06b625576b752251 along the lines of what you were thinking?

File: lib/ansible/modules/files/unarchive.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ class ZipArchive(object):
 
     def __init__(self, src, dest, file_args, module):
```
Comment (sivel) [inline lib/ansible/modules/files/unarchive.py]: Go ahead and use `b_dest` in the function signature also.

File: lib/ansible/modules/files/unarchive.py [hunk 2]
```diff
@@ -435,7 +435,7 @@ def is_unarchived(self):
             # DEBUG
 #            err += "%s%s %10d %s\n" % (ztype, permstr, size, path)
 
-            dest = os.path.join(self.dest, path)
+            dest = os.path.join(self.b_dest, to_bytes(path, errors='surrogate_or_strict'))
```
Comment (sivel) [inline lib/ansible/modules/files/unarchive.py]: Use `b_dest` as the var here.  It'll likely require more changes, but for long lived bytes, we should differentiate by using `b_` prefix

File: lib/ansible/modules/files/unarchive.py [hunk 3]
```diff
@@ -856,12 +857,12 @@ def main():
...
-    if not os.path.isdir(dest):
-        module.fail_json(msg="Destination '%s' is not a directory" % dest)
+    if not os.path.isdir(b_dest):
+        module.fail_json(msg="Destination '%s' is not a directory" % b_dest)
 
-    handler = pick_handler(src, dest, file_args, module)
+    handler = pick_handler(src, b_dest, file_args, module)
 
-    res_args = dict(handler=handler.__class__.__name__, dest=dest, src=src)
+    res_args = dict(handler=handler.__class__.__name__, dest=b_dest, src=src)
```
Comment (sivel) [inline lib/ansible/modules/files/unarchive.py]: You might save `dest` and `b_dest` as vars, as here you'll want to use `dest` as "native".  Anywhere it's going back out through `fail/exit_json` we want native.  This applies to the next several changes too.

File: lib/ansible/modules/files/unarchive.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/modules/files/unarchive.py]: This will apply to the other classes as well.

**Review by sivel — CHANGES_REQUESTED** [review]

**Review by sivel — APPROVED** [review]

**Comment by sivel** [comment]
Note that I haven't tested these changes with the inventory hack locally on freebsd/12.0 and freebsd/11.1.  I'm going to assume someone else has or will.  But visually this looks good.

File: lib/ansible/modules/files/unarchive.py (hunk 3 above)
Comment (abadger) [reply lib/ansible/modules/files/unarchive.py]: Note that fail/exit_json should be handling bytes or text; making them the proper string type for that python version before passing them in to the stdlib's json.dumps().

**Pull request merged**